index.js

```diff
@@ -41,57 +41,57 @@ module.exports = function(options) {
   options.keepHeadersOnError = options.keepHeadersOnError === undefined || !!options.keepHeadersOnError;
 
   return function cors(ctx, next) {
-    // If the Origin header is not present terminate this set of steps.
-    // The request is outside the scope of this specification.
-    const requestOrigin = ctx.get('Origin');
+    return new Promise(resolve => {
+      // If the Origin header is not present terminate this set of steps.
+      // The request is outside the scope of this specification.
+      const requestOrigin = ctx.get('Origin');
 
-    // Always set Vary header
-    // https://github.com/rs/cors/issues/10
-    ctx.vary('Origin');
+      // Always set Vary header
+      // https://github.com/rs/cors/issues/10
+      ctx.vary('Origin');
 
-    if (!requestOrigin) {
-      return next();
-    }
+      if (!requestOrigin) {
+        return resolve();
+      }
 
-    let origin;
+      if (typeof options.origin === 'function') {
+        return resolve(options.origin(ctx));
+      }
 
-    if (typeof options.origin === 'function') {
-      // FIXME: origin can be promise
-      origin = options.origin(ctx);
+      return resolve(options.origin || requestOrigin);
```

Comment:
seems it could be `return resolve(requestOrigin)`

Reply:
That'd break one test.

Reply:
@dead-horse shall we regard it as a break change and rewrite the ancient test? or do we still need to support normal function?

Reply:
we should keep support normal function as well.